### ISSUE TITLE
Better DevExp for serialization feature support as of GraalVM 21

### ISSUE THREAD

**Issue by machi1990** [quarkusio/quarkus#14530]
**Description**

The new release of [GraalVM 21.0](https://www.graalvm.org/release-notes/21_0/) adds serialization support, developers can configure class for serialization via the serialization configuration file `-H:SerializationConfigurationResources=/path/to-serialization-config.json` option. 

This release will be [supported in Quarkus soon](https://github.com/quarkusio/quarkus/pull/14512) thanks to @zakkak .

 
This can at the moment be achieved by `quarkus.native.additional-build-args=-H:SerializationConfigurationResources=/path/to-serialization-config.json`, however we can add features like the one we added for `H:ReflectionConfigurationFiles` option i.e `@RegisterForReflection` annotation for  better developer experience. 

We can also, add a way (programmaticaly) to include classes that are not part of the application e.g the ones coming from external libs / frameworks (I am wondering if there are still efforts to do this for the reflection configuration??), instead of the developer having to use the json configuration file. 

/cc @gsmet @geoand @stuartwdouglas @zakkak @emmanuelbernard

**Comment by geoand** [comment]
Thanks for bringing this up @machi1990 

Although it's interesting, I don't think anyone has ever complained about serialization not working in native, which probably means no one relies on it.
In that light, I propose we hold off from adding anything unless users really ask for it

**Comment by gsmet** [comment]
So... my secret dream would be to have a Wicket extension for Quarkus and we would need to be able to register classes for serialization for this to work :).

**Comment by geoand** [comment]
Then that make you the first user to request it 😂

**Comment by machi1990** [comment]
> Then that make you the first user to request it

I think we are not seeing the request here (yet), as this has been mostly a GraalVM issue.

**Comment by gsmet** [comment]
For me, it's the role of Quarkus to expose this and make it easier to use, even if in the end few people use it.

**Comment by geoand** [comment]
I would argue that it's probably prudent to reduce our surface by not introducing features that users don't have a need for.

**Comment by gabrielsson** [comment]
It would be great with serialization as I use hibernate. Composite keys, special db types etc then can be supported. Simple things like binding hibernate with JoinColumns without foreign keys could be utilized, so there is great potential, at least for me with support for GraalVM 21 in general, and serialization in particular.

**Comment by ppalaga** [comment]
We'd very much welcome some sort of `SerializableClassBuildItem` being delivered by Quarkus. Actually we have a bunch of issues that were originally blocked by https://github.com/oracle/graal/issues/460 and https://github.com/oracle/graal/pull/2730 now are blocked by this one :).

**Comment by JiriOndrusek** [comment]
I'd like to create SerializationClassBuildItem and @RegisterForSerialization   (both similar to reflection registration)
I'd like to create PR with contribution linked to this issue. Is there any other functionality which should be considered as a part of this feature?

**Comment by ppalaga** [comment]
> I'd like to create SerializableClassBuildItem and @RegisterForSerialization (both similar to reflection registration)
> I'd like to create PR with contribution linked to this issue. Is there any other functionality which should be considered as a part of this feature?

I volunteer to review and assist if necessary.

**Comment by gsmet** [comment]
I suspect we will also need something similar to the `ReflectiveHierarchyBuildItem`.

**Comment by gsmet** [comment]
(Except if GraalVM handles it automatically!)

**Comment by stuartwdouglas** [comment]
So I had a quick look and I think that maybe it is just an extra flag on fields to allow final fields to also be set (i.e when fields are registered for reflection, you can basically set a flag that says that final fields can also be updated).

If this is all it is then maybe it should tie into our existing reflection infrastructure.

**Comment by JiriOndrusek** [comment]
PR with the implementation is prepared and approved to be merged (https://github.com/quarkusio/quarkus/pull/15380)
